### PR TITLE
[IOS] Integrate Intro Screen logic

### PR DESCRIPTION
**Pull request by zhenyapavlovsky**
## What's new
1.  Integrate Intro Screen logic 

## Issue

[Issue](https://trello.com/c/8Jt4bDCq)

## Screenshot

Screenshot 2023-03-01 at 14 52 27

### CODE REVIEW

File: Crypto_App_iOS/Controls/ButtonView.swift [hunk 1]
```diff
@@ -8,9 +8,12 @@
 import SwiftUI
 
 struct ButtonView: View {
```
Comment (romansorochak) [inline Crypto_App_iOS/Controls/ButtonView.swift:10]: `IntroView`

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+//  CryptoCoordinator.swift
+//  Crypto_App_iOS
+//
+//  Created by Zhenya Pavlovskyi on 01.03.2023.
+//
+
+import Foundation
+import SwiftUI
+
+class CryptoCoordinator: ObservableObject {
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinator.swift:11]: `AppCoordinator`

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift [hunk 3]
```diff
@@ -19,18 +19,20 @@ struct IntroView: View {
...
-            .opacity(viewModel.showingTopTitle ? 1.0 : 0.0)
+            .opacity(viewModel.titleShown ? 1.0 : 0.0)
             .onAppear() {
-                withAnimation(Animation.easeInOut(duration: 1.0)) {viewModel.showingTopTitle.toggle()}
+                withAnimation(Animation.easeInOut(duration: 1.0)) {viewModel.titleShown.toggle()}
             }
             Spacer()
             
-                ButtonView()
+            ButtonView(viewModel: IntroViewModel(
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:28]: this looks weird 
What is happening here ??? Inside `IntroView` when user clicks on a button you create `IntroViewModel` ????
If you need to navigate to some other place on button click then you should use coordinator.

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinator.swift (hunk 2 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinator.swift:11]: Done

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinatorView.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+//  CryptoCoordinatorView.swift
+//  Crypto_App_iOS
+//
+//  Created by Zhenya Pavlovskyi on 01.03.2023.
+//
+
+import SwiftUI
+import SwiftUINavigation
+
+struct CryptoCoordinatorView: View {
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/CryptoCoordinatorView.swift:11]: Done

File: Crypto_App_iOS/Controls/ButtonView.swift (hunk 1 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Controls/ButtonView.swift:10]: Done

File: Crypto_App_iOS/Controls/ButtonView.swift [hunk 5]
```diff
@@ -8,9 +8,12 @@
...
+    
+    @ObservedObject var viewModel: IntroViewModel
+    
     var body: some View {
         
-        Button(action: {}, label: {
+        Button(action: {viewModel.startButton()}, label: {
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Controls/ButtonView.swift:16]: Done

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        case main
+        case intro
+    }
+
+    @Published var route: Route?
+
+    var mainCoordinator: MainCoordinator
+    var introViewModel: IntroViewModel
+
+    init(userDefault: UserDefaultsManager, mainCoordinator: MainCoordinator,  introViewModel: IntroViewModel) {
```
Comment (vasyl-nadtochii) [inline Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]:
```suggestion
    init(userDefault: UserDefaultsManager, mainCoordinator: MainCoordinator, introViewModel: IntroViewModel) {
```

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift [hunk 7]
```diff
@@ -19,18 +19,18 @@ struct IntroView: View {
                 toTheCryptoTitle
             }
             .padding(.top, 25)
-            .opacity(viewModel.showingTopTitle ? 1.0 : 0.0)
+            .opacity(viewModel.titleShown ? 1.0 : 0.0)
             .onAppear() {
-                withAnimation(Animation.easeInOut(duration: 1.0)) {viewModel.showingTopTitle.toggle()}
+                withAnimation(Animation.easeInOut(duration: 1.0)) {viewModel.titleShown.toggle()}
```
Comment (vasyl-nadtochii) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:24]:
```suggestion
                withAnimation(Animation.easeInOut(duration: 1.0)) { viewModel.titleShown.toggle() }
```

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift [hunk 8]
```diff
@@ -19,18 +19,18 @@ struct IntroView: View {
...
             
-                ButtonView()
+            bottomButton
             
             .padding(.bottom, 30)
-            .opacity(viewModel.showingBottomTitle ? 1.0 : 0.0)
+            .opacity(viewModel.buttonShown ? 1.0 : 0.0)
             .onAppear() {
-                withAnimation(Animation.easeInOut(duration: 2.0)) {viewModel.showingBottomTitle.toggle()}
+                withAnimation(Animation.easeInOut(duration: 2.0)) {viewModel.buttonShown.toggle()}
```
Comment (vasyl-nadtochii) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:33]:
```suggestion
                withAnimation(Animation.easeInOut(duration: 2.0)) { viewModel.buttonShown.toggle() }
```

**Review by vasyl-nadtochii — APPROVED** [review]

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift (hunk 7 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:24]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift (hunk 8 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:33]: Done

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        case main
+        case intro
+    }
+
+    @Published var route: Route?
+
+    var mainCoordinator: MainCoordinator
+    var introViewModel: IntroViewModel
+
+    init(userDefault: UserDefaultsManager, mainCoordinator: MainCoordinator,
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: `mainCoordinator: MainCoordinator,` to the next line

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift [hunk 10]
```diff
@@ -12,7 +12,9 @@ struct Crypto_App_iOSApp: App {
     
     var body: some Scene {
         WindowGroup {
-            MainCoordinatorView(coordinator: MainCoordinator())
+            NavigationView {
+                AppCoordinatorView(coordinator: .init(userDefault: UserDefaultsManager(), mainCoordinator: MainCoordinator(), introViewModel: IntroViewModel(userDefaultsManager: UserDefaultsManager())))
```
Comment (romansorochak) [inline Crypto_App_iOS/App/Crypto_App_iOSApp.swift:16]: not readable 
put each param in the new line

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 11]
```diff
@@ -9,7 +9,22 @@ import Foundation
 
 class IntroViewModel: ObservableObject {
     
-    @Published var showingTopTitle = false
-    @Published var showingBottomTitle = false
+    enum Result {
+        case main
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:13]: `main` is not a result. 
Result can be like `done` or `finished`

File: Crypto_App_iOS/Managers/UserDefaultsManager.swift [hunk 12]
```diff
@@ -0,0 +1,21 @@
+//
+//  UserDefaultsManager.swift
+//  Crypto_App_iOS
+//
+//  Created by Zhenya Pavlovskyi on 01.03.2023.
+//
+
+import Foundation
+
+class UserDefaultsManager {
```
Comment (romansorochak) [inline Crypto_App_iOS/Managers/UserDefaultsManager.swift:10]: create `protocol PropertiesManager` with the only property `introCompleted`.
Then create `class UserDefaultsManager: PropertiesManager`
This way you can use `PropertiesManager` everywhere in declarations but you will create instance of `UserDefaultsManager`.
This way you have higher level of abstraction and can actually mock this in future once you start writing unit tests for a class where you need to use `PropertiesManager`.

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift (hunk 9 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: Done

Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: Done

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift (hunk 10 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/App/Crypto_App_iOSApp.swift:16]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift (hunk 11 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:13]: Done

File: Crypto_App_iOS/Managers/UserDefaultsManager.swift (hunk 12 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Managers/UserDefaultsManager.swift:10]: Done

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift [hunk 13]
```diff
@@ -12,7 +12,11 @@ struct Crypto_App_iOSApp: App {
     
     var body: some Scene {
         WindowGroup {
-            MainCoordinatorView(coordinator: MainCoordinator())
+            NavigationView {
+                AppCoordinatorView(coordinator: .init(userDefault: UserDefaultsManager()
```
Comment (romansorochak) [inline Crypto_App_iOS/App/Crypto_App_iOSApp.swift:16]: move `coordinator` param to the next line

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift [hunk 14]
```diff
@@ -12,7 +12,11 @@ struct Crypto_App_iOSApp: App {
...
+                                                      , mainCoordinator: MainCoordinator()
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/App/Crypto_App_iOSApp.swift:17]: Done

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift (hunk 13 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/App/Crypto_App_iOSApp.swift:16]: Done

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift [hunk 15]
```diff
@@ -12,7 +12,11 @@ struct Crypto_App_iOSApp: App {
...
+                                                      , introViewModel: IntroViewModel(userDefaultsManager: UserDefaultsManager())))
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/App/Crypto_App_iOSApp.swift:18]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 16]
```diff
@@ -9,7 +9,22 @@ import Foundation
...
-    @Published var showingBottomTitle = false
+    enum Result {
+        case onMainSelected
+    }
     
+    @Published var titleShown = false
+    @Published var buttonShown = false
+    
+    var onResult: ((Result) -> Void)?
+    var userDefaultsManager: UserDefaultsManager
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:20]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 17]
```diff
@@ -9,7 +9,22 @@ import Foundation
...
+    
+    init(userDefaultsManager: UserDefaultsManager) {
+        self.userDefaultsManager = userDefaultsManager
+    }
+    
+    func startButton() {
+        onResult?(.onMainSelected)
+        userDefaultsManager.introCompleted = true
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:28]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 18]
```diff
@@ -9,7 +9,22 @@ import Foundation
...
+        case onMainSelected
+    }
     
+    @Published var titleShown = false
+    @Published var buttonShown = false
+    
+    var onResult: ((Result) -> Void)?
+    var userDefaultsManager: UserDefaultsManager
+    
+    init(userDefaultsManager: UserDefaultsManager) {
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:22]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift (hunk 17 above)
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:28]: Done

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+        case main
+        case intro
+    }
+
+    @Published var route: Route?
+
+    var mainCoordinator: MainCoordinator
+    var introViewModel: IntroViewModel
+
+    init(userDefault: UserDefaultsManager,
```
[the remark this answers is not in the document]
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: Done

File: Crypto_App_iOS/App/Crypto_App_iOSApp.swift [hunk 20]
```diff
@@ -12,7 +12,14 @@ struct Crypto_App_iOSApp: App {
...
     var body: some Scene {
         WindowGroup {
-            MainCoordinatorView(coordinator: MainCoordinator())
+            NavigationView {
+                AppCoordinatorView(coordinator: .init(
+                    propertiesManager: PreferencesImpl(),
+                    mainCoordinator: MainCoordinator(),
+                    introViewModel: IntroViewModel(propertiesManager: PreferencesImpl())
+                )
+                )
```
Comment (romansorochak) [inline Crypto_App_iOS/App/Crypto_App_iOSApp.swift:21]: bad alignment I believe

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        case main
+        case intro
+    }
+
+    @Published var route: Route?
+
+    var mainCoordinator: MainCoordinator
+    var introViewModel: IntroViewModel
+
+    init(propertiesManager: PreferencesImpl,
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: `init(preferences: Preferences,`

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 22]
```diff
@@ -9,7 +9,22 @@ import Foundation
...
+        case onMainSelected
+    }
     
+    @Published var titleShown = false
+    @Published var buttonShown = false
+    
+    var onResult: ((Result) -> Void)?
+    var propertiesManager: PreferencesImpl
+    
+    init(propertiesManager: PreferencesImpl) {
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:22]: `init(preferences: Preferences) {`

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift [hunk 23]
```diff
@@ -9,7 +9,22 @@ import Foundation
...
+        self.propertiesManager = propertiesManager
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:23]:
```suggestion
        self.preferences = preferences
```

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift [hunk 24]
```diff
@@ -19,24 +19,26 @@ struct IntroView: View {
                 toTheCryptoTitle
             }
             .padding(.top, 25)
-            .opacity(viewModel.showingTopTitle ? 1.0 : 0.0)
+            .opacity(viewModel.titleShown ? 1.0 : 0.0)
             .onAppear() {
-                withAnimation(Animation.easeInOut(duration: 1.0)) {viewModel.showingTopTitle.toggle()}
+                withAnimation(Animation.easeInOut(duration: 1.0))
+                {viewModel.titleShown.toggle()}
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:25]: why `{` is in this line ?
it must always be in the previous line

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift [hunk 25]
```diff
@@ -19,24 +19,26 @@ struct IntroView: View {
...
-            .padding(.bottom, 30)
-            .opacity(viewModel.showingBottomTitle ? 1.0 : 0.0)
-            .onAppear() {
-                withAnimation(Animation.easeInOut(duration: 2.0)) {viewModel.showingBottomTitle.toggle()}
-            }
+                .padding(.bottom, 30)
+                .opacity(viewModel.buttonShown ? 1.0 : 0.0)
+                .onAppear() {
+                    withAnimation(Animation.easeInOut(duration: 2.0))
+                    {viewModel.buttonShown.toggle()}
```
Comment (romansorochak) [inline Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:35]: same

File: Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift (hunk 21 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Coordination/CryptoApp/AppCoordinator.swift:23]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift (hunk 23 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:23]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift (hunk 22 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroViewModel.swift:22]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift (hunk 24 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:25]: Done

File: Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift (hunk 25 above)
Comment (zhenyapavlovsky) [reply Crypto_App_iOS/Flows/Screens/Intro/IntroView.swift:35]: Done

**Review by romansorochak — APPROVED** [review]

**Pull request merged**